File: src/tmp/Python/string/zigzag_conversion.py
def convert(s: str, numRows: int) -> str:
    if numRows == 1:
        return s
    
    row = min(len(s), numRows)
    remain = [[] for i in range(row)]
    r, flag = 0, False
    for i, ch in enumerate(s):
        remain[r].append(ch)
        if r == 0 or r == row - 1:
            flag = not flag
        if flag:
            r += 1
        else:
            r -= 1
    result = []
    for i in range(row):
        result.extend(remain[i])
    return "".join(result)

if __name__ == "__main__":
    print(convert(s = "A", numRows = 1))

File: core/tmp/Java/src/com/dbc/algorithm/String/ZigzagConversion.java
package com.dbc.algorithm.String;

import java.util.ArrayList;
import java.util.List;

public class ZigzagConversion {
    public static String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }
        int row = Math.min(numRows, s.length());
        List<StringBuffer> remain = new ArrayList<>();
        for (int i = 0; i < row; i++) {
            remain.add(new StringBuffer());
        }

        int goDown = 0;
        boolean flag = false;
        for (int i = 0; i < s.length(); i++) {
            remain.get(goDown).append(s.charAt(i));
            if (goDown == 0 || goDown == row - 1) {
                flag = !flag;
            }
            goDown += flag ? 1 : -1;
        }
        StringBuilder result = new StringBuilder();
        for (StringBuffer buffer : remain) {
            result.append(buffer);
        }
        return result.toString();
    }
}
